File: spring/src/main/java/com/okaka/onenightjinroh/api/GetGameRuleController.java
package com.okaka.onenightjinroh.api;

import com.okaka.jinroh.persistence.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.servlet.http.HttpSession;
import java.util.List;
import java.util.stream.Collectors;

@RestController
public class GetGameRuleController {
    @Autowired
    HttpSession session;

    @Autowired
    GameDao gameDao;

    @Autowired
    RoleSelectDao roleSelectDao;

    @Autowired
    RoleDao roleDao;

    @RequestMapping(path = "/game-rule/{gameId}")
    Response getGameRule(@PathVariable Long gameId) {
        GameEntity game = gameDao.select(gameId);
        Long ruleId = game.rule_id;

        List<RoleEntity> gameRuleRoleList = roleSelectDao.selectRoleListByRuleId(ruleId);

        List<RoleEntity> roleList = roleDao.selectAll();

        List<RoleResponse> roleResponseList = roleList.stream().map(role ->
                new RoleResponse(
                        role.role_id,
                        role.role_name,
                        gameRuleRoleList.stream().filter(it -> it.role_id.equals(role.role_id)).count()
                )
        ).collect(Collectors.toUnmodifiableList());

        return new Response(roleResponseList);
    }

    public class Response{
        List<RoleResponse> roleList;

        public Response(List<RoleResponse> roleList){
            this.roleList = roleList;
        }

        public List<RoleResponse> getRoleList() {
            return roleList;
        }
    }

    public class RoleResponse{
        Long roleId;
        String roleName;
        Long count;

        public RoleResponse(
            Long roleId,
            String roleName,
            Long count
        ){
            this.roleId = roleId;
            this.roleName = roleName;
            this.count = count;

        }

        public Long getRoleId() {
            return roleId;
        }

        public String getRoleName() {
            return roleName;
        }

        public Long getCount() {
            return count;
        }
    }
}

File: spring/src/main/java/com/okaka/onenightjinroh/api/GetGameRuleController.kt
package com.okaka.onenightjinroh.api

import com.okaka.jinroh.persistence.GameDao
import com.okaka.jinroh.persistence.RoleDao
import com.okaka.jinroh.persistence.RoleEntity
import com.okaka.jinroh.persistence.RoleSelectDao
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
class GetGameRuleController(
    private val gameDao: GameDao,
    private val roleSelectDao: RoleSelectDao,
    private val roleDao: RoleDao
) {
    @RequestMapping(path = ["/game-rule/{gameId}"])
    fun getGameRule(@PathVariable gameId: Long?): Response {
        val game = gameDao.select(gameId)
        val gameRuleRoleList = roleSelectDao.selectRoleListByRuleId(game.rule_id)
        val roleList = roleDao.selectAll()
        val roleResponseList = roleList.map { role: RoleEntity ->
            RoleResponse(
                role.role_id,
                role.role_name,
                gameRuleRoleList.count { it.role_id == role.role_id }
            )
        }

        return Response(roleResponseList)
    }

    class Response(
        val roleList: List<RoleResponse>
    )

    class RoleResponse(
        val roleId: Long,
        val roleName: String,
        val count: Int
    )
}
